pravega/pravega
# Issue 4352: Turn retrieve token operation of DelegationTokenProvider asynchronous

**Change log description**  
Make the `DelegationTokenProvider` interface operation `retrieveToken()` asynchronous and make its invocations in client non-blocking, in order to avoid client-side deadlocks. 

**Purpose of the change**  
Fixes #4352 

**What the code does**  
* Makes the `retrieveToken()` operation of the `DelegationTokenProvider` interface asynchronous, and modifies the following implementation classes accordingly: 
  * `JwtTokenProviderImpl` 
  * `EmptyTokenProviderImpl` 
  * `StringTokenProviderImpl`
* Modifies the following client classes to avoid blocking, by leveraging `CompletableFuture`'s `CompletionStage` tasks: 
  * `AsyncSegmentInputStreamImpl`
  * `ConditionalOutputStreamImpl.java`
  * `SegmentMetadataClientImpl.java`
  * `SegmentOutputStreamImpl.java`
* Modifies existing tests to account for the above items. 

* Modifies existing code 

**How to verify it**  
All existing tests - unit, integration and system tests - should pass. `Auth` enabled autoscale system tests should pass.  

Here's a subset of unit and integration tests I use to quickly verify the changes: 
```java
package io.pravega.test.integration;

import io.pravega.client.security.auth.DelegationTokenProviderFactoryTest;
import io.pravega.client.security.auth.JwtBodyTest;
import io.pravega.client.security.auth.JwtTokenProviderImplTest;
import io.pravega.client.segment.impl.AsyncSegmentInputStreamTest;
import io.pravega.client.segment.impl.ConditionalOutputStreamTest;
import io.pravega.client.segment.impl.SegmentMetadataClientTest;
import io.pravega.client.segment.impl.SegmentOutputStreamTest;
import io.pravega.client.state.impl.RevisionedStreamClientTest;
import io.pravega.client.stream.impl.EventStreamReaderTest;
import io.pravega.client.stream.impl.SegmentSelectorTest;
import io.pravega.segmentstore.server.host.handler.PravegaRequestProcessorAuthFailedTest;
import io.pravega.shared.protocol.netty.WireCommandsTest;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;

@RunWith(Suite.class)
@Suite.SuiteClasses({
        JwtTokenProviderImplTest.class,
        DelegationTokenProviderFactoryTest.class,
        JwtBodyTest.class,
        EventStreamReaderTest.class,
        RevisionedStreamClientTest.class,
        AppendTest.class,
        AppendReconnectTest.class,
        ReadTest.class,
        SegmentSelectorTest.class,
        ConditionalOutputStreamTest.class,
        PravegaRequestProcessorAuthFailedTest.class,
        WireCommandsTest.class,
        AsyncSegmentInputStreamTest.class,
        SegmentMetadataClientTest.class,
        SegmentOutputStreamTest.class,
        DelegationTokenTest.class
})

public class DelegationTokenTestSuite {
}
```
